### PR TITLE
Better jupyter support

### PR DESCRIPTION
**Pull request by MatthewWilkes**
## Description
This adds a patch for ipython's IO streams and wraps subprocess calls to make them more tolerant of the fact that jupyter does not allocate file descriptors for its streams. This allows experiments to be run through jupyter, albeit with the caveat that output is significantly laggier than using a console script.

## Motivation and Context
This allows the full range of functionality to be used in a Jupyter notebook, lowering the bar to entry.

## How Has This Been Tested?
Manual tests of bartlett1932/demo.ipynb only.

### CODE REVIEW

**Comment by jessesnyder** [comment]
@MatthewWilkes Is the lagginess you refer to only in the context of Jupyter?

File: dallinger/patches.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+def patch_ipython():
+    try:
+        from ipykernel.iostream import OutStream
+    except ImportError:
+        return
+    else:
+        OutStream.writable = lambda self: True
+
+
+def patch():
```
Comment (jessesnyder) [inline dallinger/patches.py]: Why the alias? (No implied critique; genuinely curious.)

**Comment by MatthewWilkes** [comment]
Sounds good, looks like I missed something in the rebase too, will look at that today.

File: dallinger/patches.py (hunk 1 above)
Comment (MatthewWilkes) [reply dallinger/patches.py]: Just to stop it filling up `__init__.py` with imports if more patches get added in future.

**Comment by MatthewWilkes** [comment]
@jessesnyder Yep, the lagginess is a Jupyter only thing. If it detects that `stdout` or `stderr` aren't real files it allocates a temporary file and hooks that up as the output of the subprocess. Once that finishes it copies the stuff from the tempfile to `sys.stdout` or `sys.stderr`, which is the connection to the Jupyter socket.

If not running under Jupyter then stdout will be a real file, so it will pass through without the buffering.

**Pull request merged**